fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,139 @@
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    const headers = {
+      'Content-Type': 'application/json'
+    };
+    fetch('/api/todos', {
+      method: 'POST',
+      headers,
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(data => {
+        const todosCopy = this.state.todos;
+        const joinedTodos = todosCopy.concat(data);
+        this.setState({ todos: joinedTodos });
+      })
+      .catch(err => console.error(err));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    let toggledTodoIndex = null;
+
+    const todo = { isCompleted: null };
+
+    for (var i = 0; i < this.state.todos.length; i++) {
+      if (this.state.todos[i].todoId === todoId) {
+        toggledTodoIndex = i;
+      }
+    }
+
+    if (this.state.todos[toggledTodoIndex].isCompleted) {
+      todo.isCompleted = false;
+    } else {
+      todo.isCompleted = true;
+    }
+
+    const headers = {
+      'Content-Type': 'application/json'
+    };
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers,
+      body: JSON.stringify(todo)
+    }
+    )
+      .then(response => {
+        return response.json();
+      })
+      .then(data => {
+        const todosCopy = this.state.todos;
+        todosCopy[toggledTodoIndex] = data;
+        this.setState({ todos: todosCopy });
```

Comment:
To make a legit copy you can use `.slice()` or `...` (spread)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#description

Reply:
on it